### PR TITLE
Refactor Java remote API Command hierarchy

### PR DESCRIPTION
**Pull request by labkey-adam**
#### Rationale
Refactor `Command` hierarchy for consistency. Other fixes and enhancements. See [5.0.0 change log](https://github.com/LabKey/labkey-api-java/blob/2d3370c0b1c327a2cf490614ddedaa1424e49772/CHANGELOG.md) for details.

### CODE REVIEW

File: src/org/labkey/remoteapi/CommandResponse.java [hunk 1]
```diff
@@ -46,16 +46,16 @@
      * @param text The response text
      * @param statusCode The HTTP status code
      * @param contentType The response content type
-     * @param json The parsed JSONObject (or null if JSON was not returned).
-     * @param sourceCommand A copy of the command that created this response
+     * @param json The parsed JSONObject (or null if JSON was not returned)
+     * @param hasRequiredVersion An object that implements HasRequiredVersion, such as the command that created this response
      */
-    public CommandResponse(String text, int statusCode, String contentType, JSONObject json, Command<?> sourceCommand)
+    public CommandResponse(String text, int statusCode, String contentType, JSONObject json, HasRequiredVersion hasRequiredVersion)
```
Comment (labkey-tchad) [inline src/org/labkey/remoteapi/CommandResponse.java]: Can we just remove api version handling from the base response class? There's only one call to `getRequiredVersion` across our entire code base. Seems strange to have this vestigial parameter on every response constructor. The response classes that care about the api version can include it without having to pass it up to the base response.

File: src/org/labkey/remoteapi/Command.java [hunk 2]
```diff
@@ -129,26 +108,13 @@ public String getActionName()
...
-    }
-
-    /**
-     * Sets the current parameter map.
-     * @param parameters The parameter map to use
-     */
-    public void setParameters(Map<String, Object> parameters)
-    {
-        _parameters = parameters;
+        return new HashMap<>();
```
Comment (labkey-tchad) [inline src/org/labkey/remoteapi/Command.java:128]: Should we use an unmodifiable map to fail hard if people try to use the old `getParameters().put(..)` pattern?
```suggestion
        return Collections.emptyMap();
```

`Command.createURI` would need to deal with this and subclasses would need to not build off of `super.getParameters`.

File: src/org/labkey/remoteapi/SimpleGetCommand.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        _parameterMapFactory = new Supplier<>()
+        {
+            private final Map<String, Object> _parameters = new HashMap<>(parameters);
+
+            @Override
+            public Map<String, Object> get()
+            {
+                return new HashMap<>(_parameters);
+            }
```
Comment (labkey-tchad) [inline src/org/labkey/remoteapi/SimpleGetCommand.java:41]: Make unmodifiable?
```suggestion
            public Map<String, Object> get()
            {
                return Collections.unmodifiableMap(_parameters);
            }
```

File: src/org/labkey/remoteapi/SimplePostCommand.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        super(controllerName, actionName);
+    }
+    /**
+     * Returns the JSON object to post or null for no JSON.
+     * @return The JSON object to post.
+     */
+
+    @Override
+    public JSONObject getJsonObject()
```
Comment (labkey-tchad) [inline src/org/labkey/remoteapi/SimplePostCommand.java]: Fix spacing
```suggestion
    }

    /**
     * Returns the JSON object to post or null for no JSON.
     * @return The JSON object to post.
     */
    @Override
    public JSONObject getJsonObject()
```

File: src/org/labkey/remoteapi/query/SelectRowsResponse.java [hunk 5]
```diff
@@ -66,7 +66,7 @@ else if("date".equalsIgnoreCase(type))
      * @param json The parsed JSONObject (or null if no JSON was returned
      * @param sourceCommand A copy of the command that created this response
      */
-    public SelectRowsResponse(String text, int statusCode, String contentType, JSONObject json, Command<? extends SelectRowsResponse> sourceCommand)
+    public SelectRowsResponse(String text, int statusCode, String contentType, JSONObject json, Command<? extends SelectRowsResponse, ?> sourceCommand)
     {
         super(text, statusCode, contentType, json, sourceCommand);
     }
```
Comment (labkey-tchad) [inline src/org/labkey/remoteapi/query/SelectRowsResponse.java:72]:
```suggestion
    public SelectRowsResponse(String text, int statusCode, String contentType, JSONObject json, HasRequiredVersion hasRequiredVersion)
    {
        super(text, statusCode, contentType, json, hasRequiredVersion);
    }
```

File: src/org/labkey/remoteapi/CommandResponse.java (hunk 1 above)
Comment (labkey-adam) [reply src/org/labkey/remoteapi/CommandResponse.java]: Sure, sounds good

File: src/org/labkey/remoteapi/Command.java (hunk 2 above)
Comment (labkey-adam) [reply src/org/labkey/remoteapi/Command.java:128]: `protected createParameterMap()` returns a mutable Map. `public final getParameters()` now returns an immutable copy.

File: src/org/labkey/remoteapi/SimpleGetCommand.java (hunk 3 above)
Comment (labkey-adam) [reply src/org/labkey/remoteapi/SimpleGetCommand.java:41]: Addressed by `getParameters()`/`createParameterMap()` change. This is now returned by `createParameterMap()`, which is `protected` and advertised as returning a mutable `Map`.

File: src/org/labkey/remoteapi/query/SelectRowsResponse.java (hunk 5 above)
Comment (labkey-adam) [reply src/org/labkey/remoteapi/query/SelectRowsResponse.java:72]: Done

File: src/org/labkey/remoteapi/SimplePostCommand.java (hunk 4 above)
Comment (labkey-adam) [reply src/org/labkey/remoteapi/SimplePostCommand.java]: Done

File: src/org/labkey/remoteapi/Command.java [hunk 6]
```diff
@@ -107,12 +108,22 @@ public String getActionName()
...
+        return Collections.unmodifiableMap(createParameterMap());
+    }
+
     /**
      * Returns a new, mutable parameter map. Derived classes will typically override this
      * method to put values passed to specialized setter methods into the map.
      * @return The parameter map to use when building the URL.
      */
-    public Map<String, Object> getParameters()
+    protected Map<String, Object> createParameterMap()
```
Comment (labkey-tchad) [inline src/org/labkey/remoteapi/Command.java:126]: Having two methods to get parameters seems like a recipe for confusion as to which one should be overridden by subclasses. What if we made `getParameters` final?

Comment (labkey-adam) [reply src/org/labkey/remoteapi/Command.java:126]: Yes, I meant to do that. Good catch.

**Review by labkey-tchad — APPROVED** [review]

**Pull request merged**